### ISSUE TITLE
Support Windows Containers

### ISSUE THREAD

**Issue by seanyoung** [containers/podman#8136]
I realise this might be out of scope but I thought I'd raise it so it can be considered.

Windows Containers are slightly modified docker containers, which contain windows executables and the windows registry. These file systems could be mounted with case-insensitive options, and then a very thin wine shim could run the actual binaries. From what I can see, all the necessary executables and dlls are included in the images (e.g. cmd.exe).

https://docs.microsoft.com/en-us/virtualization/windowscontainers/

This would be super-useful to be able to run e.g. windows builds on linux, see and example of a windows dockerfile here: https://github.com/hyperledger-labs/solang/blob/master/scripts/build-llvm-windows.dockerfile

**Comment by rhatdan** [comment]
This would definitely would require community to work on.  Not something the core developers are likely to work on.

**Comment by trevor-viljoen** [comment]
As much as I don't want to deal with Windows Containers, we have many Windows application workloads in the pipeline that we need to containerize. I would like to use Quay's builders to facilitate some of this work. However, until podman/buildah/libpod support this it won't be possible.

**Comment by rhatdan** [comment]
No progress and no volunteers.

**Comment by vrothberg** [comment]
I am closing this issue in favor of https://github.com/containers/podman/issues/8452

**Comment by seanyoung** [comment]
#8452 is about running podman on windows/mac. This issue about running windows containers on podman (on linux). Not the same issue.

**Comment by vrothberg** [comment]
Thanks, @seanyoung! I missed that you want to run Windows containers on Linux.

**Comment by rhatdan** [comment]
This would require something like kata to run qemu within a container.

**Comment by seanyoung** [comment]
@rhatdan no this would not require qemu. It would require a wine. I imagine there would be base layer with wine, and then the windows image as /c/. Then wine runs `/c/windows/system32/cmd.exe` (or whatever is specified as the startup executable.

The wine can be stripped down version as no gui components are required. Windows doesn't allow gui things to run in windows containers either.

**Comment by rhatdan** [comment]
@seanyoung crun has the ability to load shared libraries when running containers, we are doing something with this when setting up libkrun.  If we had a similar library around wine, we might be able to pull this off.  @giuseppe WDYT?

Then we could launch crun with a wrun OCI Link.

@seanyoung Would this be something you would be interested in working on?

**Comment by seanyoung** [comment]
@rhatdan this is something I'd like to work on, but realistically I don't know when I'll find the time. I'm already working on other open source projects. Not enough time in the day is one of life's greatest disappointments.

Having said that, I had started working on making
```
$ podman pull mcr.microsoft.com/windows/servercore:ltsc2019
Trying to pull mcr.microsoft.com/windows/servercore:ltsc2019...
  no image found in manifest list for architecture amd64, variant "", OS linux
Error: Error choosing an image from manifest list docker://mcr.microsoft.com/windows/servercore:ltsc2019: no image found in manifest list for architecture amd64, variant "", OS linux
```
work, but this required removing some assumptions across the code.  Should pull have a `--windows` option?

**Comment by seanyoung** [comment]
Hmm looks like `podman pull` now has `--os` and `--platform` options. It's been a while since I looked at this

**Comment by rhatdan** [comment]
$ podman pull --os windows mcr.microsoft.com/windows/servercore:ltsc2019
Trying to pull mcr.microsoft.com/windows/servercore:ltsc2019...
Getting image source signatures
Copying blob 5ff1512f88ec [=>------------------------------------] 36.6MiB / 708.8MiB
Copying blob 4612f6d0b889 [--------------------------------------] 1.3MiB / 1.6GiB
...

**Comment by rhatdan** [comment]
Sadly this failed. Not sure why.
```
 $ podman pull --os windows mcr.microsoft.com/windows/servercore:ltsc2019
Trying to pull mcr.microsoft.com/windows/servercore:ltsc2019...
Getting image source signatures
Copying blob 5ff1512f88ec [================>---------------------] 316.2MiB / 708.8MiB
Copying blob 5ff1512f88ec done  
Copying blob 4612f6d0b889 done  
Copying config 3eaa9ebbf5 done  
Writing manifest to image destination
Storing signatures
  Error processing tar file(exit status 1): link /Files/Program Files/common files/Microsoft Shared/Ink/en-US/micaut.dll.mui /Files/Program Files (x86)/common files/Microsoft Shared/ink/en-US/micaut.dll.mui: no such file or directory
Error: Error committing the finished image: error adding layer with blob "sha256:5ff1512f88ecd4ff0a91ade34d43733c48834433e35f062f80e6e70d706f8d73": Error processing tar file(exit status 1): link /Files/Program Files/common files/Microsoft Shared/Ink/en-US/micaut.dll.mui /Files/Program Files (x86)/common files/Microsoft Shared/ink/en-US/micaut.dll.mui: no such file or directory
```

**Comment by seanyoung** [comment]
Yes, file links seem to work differently. I haven't quite figured out what the difference is.

The image also includes a copy of the windows registry.

Lastly the filesystem is case-insensitive. The linux kernel has support for this, but I'm not quite sure what modifications podman needs to set this up.

**Comment by rhatdan** [comment]
@rhvgoyal WDYT?

**Comment by giuseppe** [comment]
> @seanyoung crun has the ability to load shared libraries when running containers, we are doing something with this when setting up libkrun. If we had a similar library around wine, we might be able to pull this off. @giuseppe WDYT?

that is something we could add.  Is it possible to load wine from a `.so` so we don't have to bind mount external files into the container?  Does it require to load other files (that won't be accessible from the container) in order to work?

**Comment by giuseppe** [comment]
> Lastly the filesystem is case-insensitive. The linux kernel has support for this, but I'm not quite sure what modifications podman needs to set this up.

it may need to setup a different storage on a case insensitive file system (ext4 and xfs should be fine).  That will be a problem if we want the linux and windows containers to use the same storage.

**Comment by seanyoung** [comment]
It might be possible to use libwine.so so we don't have to bind mount external files. I think to get things up a running more quickly, we should first work with a bind mount. Also, since this hasn't been attempted yet, we don't know what hurdles we are going to face.

**Comment by rhatdan** [comment]
@giuseppe Any chance you could create a spike to see if we could support this?

**Comment by giuseppe** [comment]
created one: https://issues.redhat.com/browse/RUN-1262

**Comment by rhatdan** [comment]
I would love to see this move forward.

**Comment by flouthoc** [comment]
@trevor-viljoen @seanyoung @rhatdan @giuseppe From the issue it seems more like `windows-binary-on-linux` / `wine-inside-container` something like this as base image should solve the use-case right ? 

Sorry i might have missed it or was not able to infer from discussions in the above thread but could someone help me understand the use-case where runtime explicitly needs to handle `wine`.
 
```dockerfile
FROM ubuntu:18.04

# Install wget
RUN apt-get update
RUN apt-get install -y wget

# Add 32-bit architecture
RUN dpkg --add-architecture i386
RUN apt-get update

# Install Wine
RUN apt-get install -y software-properties-common gnupg2
RUN wget -nc https://dl.winehq.org/wine-builds/winehq.key
RUN apt-key add winehq.key
RUN apt-add-repository 'deb https://dl.winehq.org/wine-builds/ubuntu/ bionic main'
RUN apt-get install -y --install-recommends winehq-stable winbind

# Turn off Fixme warnings
ENV WINEDEBUG=fixme-all

# Setup a Wine prefix
ENV WINEPREFIX=/root/.net 
ENV WINEARCH=win64
RUN winecfg

# Install Winetricks
RUN apt-get install -y cabextract
RUN wget https://raw.githubusercontent.com/Winetricks/winetricks/master/src/winetricks
RUN chmod +x winetricks
RUN cp winetricks /usr/local/bin

# Install .NET Framework 4.5.2
RUN wineboot -u && winetricks -q dotnet452

# Compile your binary here

# Run Your binary
ENTRYPOINT ["wine", "<your-binary>"]```

**Comment by seanyoung** [comment]
@flouthoc the aim is not to run a single windows binary. The aim is to download and run windows containers.

Having said that, a base image with wine seems like a good way start.

**Comment by seanyoung** [comment]
@flouthoc I think a wine base image is a great way to start this.

There might be advantages to having the runtime explicitly use libwine rather than a wine base image. It might be easier to configure certain wine settings or capture wine logs, if needed. Also using libwine rather than a wine base image feels a lot more robust/native.

Having said all that, presumably using a wine base image will be easier to get this thing working and then later we can consider using libwine.

**Comment by rhatdan** [comment]
I prefer a ociwine for this.  Which is just crun linked against libwine.
The bigger problem is the file system issue.  Basically Podman or other tools would need a way to have a Windows storage directory which is setup on a different mount point.  The windows storage would require the file system to ignore case differences, so that the Wine APPS within the container would work correctly.

**Comment by rhatdan** [comment]
@flouthoc @giuseppe I would like to setup a meeting to discuss this and figure out if it is possible or close this issue.

**Comment by carmo-evan** [comment]
I think this becomes even more relevant now with Docker Desktop becoming a paid service.

**Comment by seanyoung** [comment]
Also docker for linux can't run windows containers. This is a real killer feature

**Comment by flouthoc** [comment]
Working on a PoC for this. Starting this with bind mount we can switch to linked .so after that.

**Comment by geoffreysmith** [comment]
@flouthoc I'm following the progress in multiple repos with my own repo (https://github.com/geoffreysmith/boring-euclid) ... seems to be a lot of people saying there's Windows support but not meaning running and building Windows containers. Would be interesting in helping with the development in anyway I can. Is there a discord or something?

**Comment by giuseppe** [comment]
@flouthoc are you still working on this issue?

@geoffreysmith we can discuss on #podman (irc.freenode.net), the idea is to take advantage of Wine and embed it in crun to be able to run Windows apps.  We have not managed yet since Wine uses a lot of shared libraries.  The first step is probably to get a static build of Wine

**Comment by vrothberg** [comment]
Note that the official IRC channel is on https://libera.chat/ now.

**Comment by flouthoc** [comment]
@geoffreysmith @seanyoung  I think wine might not be a good fit for this use case since running static wine in itself is not very well supported officially https://forum.winehq.org/viewtopic.php?t=31388 and the amount of shared libraries needed by wine is making plumbing and build extremely difficult for crun.

Also afaik wine has no intention of being used as an embeddable library thus making implementation extremely difficult for such use-cases.

**Comment by giuseppe** [comment]
> Note that the official IRC channel is on https://libera.chat/ now.



> Note that the official IRC channel is on https://libera.chat/ now.

thanks!  Fixed my comment

**Comment by rhatdan** [comment]
@flouthoc @giuseppe no work lately correct?

**Comment by flouthoc** [comment]
@rhatdan Doing some research again. Hopefully will update here soon.

**Comment by flouthoc** [comment]
A variant of this is being implemented for crun so I am removing stale.

**Comment by seanyoung** [comment]
@flouthoc I guess you mean https://github.com/containers/crun/pull/841 ?

How is that going to help towarrds running windows containers? This is just loading dotnet applications which can already run fine in linux containers.

**Comment by flouthoc** [comment]
@seanyoung It will be much easier to do similar loading with wine shared libraries as well but again wine has a dependency on huge set of shared libaries. But the idea of above PR is to run executables without runtime layer container image but will be only scoped to `.NET`.

**Comment by flouthoc** [comment]
@seanyoung Also the idea is pretty much same cause even `wine` works on linux if we are good to adding another extra layer in image which similar to what we do for `.NET` runtime.

**Comment by seanyoung** [comment]
> @seanyoung It will be much easier to do similar loading with wine shared libraries as well but again wine has a dependency on huge set of shared libaries. But the idea of above PR is to run executables without runtime layer in container image but will be only scoped to `.NET`.

Ok.

> Edit: I am hoping a similar bind mount for all the shared libaries for wine should make it work just like we are doing for mono. Since static building of wine is not at all easy. https://forum.winehq.org/viewtopic.php?t=31388

Windows Containers come with all the dlls included. It possible that all that is needed is an pe exe loader and the rest of the windows system calls, and the rest of wine is not needed.

**Comment by flouthoc** [comment]
> Windows Containers come with all the dlls included. It possible that all that is needed is an pe exe loader and the rest of the windows system calls, and the rest of wine is not needed.

Could be but wine loads some of its own stuff dynamically as well. I guess that's why people are exploring `static` build for wine but no one was able to get it working i guess.  Once I complete the `mono` poc I guess i should be able to get more idea.

**Comment by rhatdan** [comment]
The other big issue is file system issues.  You have to have the file that stores your images support treating files with different cases the same.  Basically running in Windows mode.  Even if we have this, I think volumes might end up being an issue for the same reason.
BTW I would love to have something like this, but we could use community help.

**Comment by apollocreed** [comment]
Hello ! 
It's a related issue, I'm trying to build a windows container with podman on WSL2 without success.
I've tried both ways.
~~~bash
podman build  --platform windows/amd64 -t dockeragent:latest  .
podman build --os windows -t dockeragent:latest  .
~~~
my Dockerfile:

~~~Dockerfile
FROM mcr.microsoft.com/windows/servercore:ltsc2019

WORKDIR /azp

COPY start.ps1 .

CMD powershell .\start.ps1
~~~

here the error 
```
Error: error creating build container: writing blob: adding layer with blob "sha256:8b534d64a7c337eb8a23b425e4f598cd3e7407ddf8c7b2f714d1e7f7ed6a04be": Error processing tar file(exit status 1): link /Files/Program Files/common files/Microsoft Shared/Ink/en-US/micaut.dll.mui /Files/Program Files (x86)/common files/Microsoft Shared/ink/en-US/micaut.dll.mui: no such file or directory

```
Seems that a DLL is missing 

my env
```
uname -a
Linux  5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
Cheers

**Comment by bb010g** [comment]
Is this also the issue for running Windows containers on Windows, and not just running Windows containers on Linux?

**Comment by flouthoc** [comment]
@bb010g I think this issues is more of a spike for running `windows containers on linux` and not `windows containers on windows` not sure but I don't think we have any open issue for `windows container on windows`.

@apollocreed For building windows containers we have a issue opened at buildah https://github.com/containers/podman/issues/8136#issuecomment-1142194900 and not yet merged PR on buildah ( working for ADD cases ) https://github.com/containers/buildah/pull/4053

**Comment by rhatdan** [comment]
There seems to be a slight effort toward this goal going on.  Some of the effort could be helped by the freebsd effort ongoing.

**Comment by lippertmarkus** [comment]
@rhatdan do you have a link regarding the ongoing effort?

**Comment by rhatdan** [comment]
https://github.com/containers/podman/pull/15372
https://github.com/containers/podman/pulls?q=is%3Aopen+is%3Apr+label%3Awindows

**Comment by lippertmarkus** [comment]
@rhatdan sorry, but can you clarify? Most issues seem to be around running a podman machine in WSL and only using the CLI on Windows. This issue here is about running Windows containers on a Linux host, so completely different, or am I missing something?

**Comment by rhatdan** [comment]
Sure, the basic premise is getting the podman commands easier to run on Windows.  The difficult part would be getting containers/storge to work, since I don't believe Windows has any support for layered file systems.  But baby steps.

As far as running Windows containers on a Linux Box, the biggest problem is handling the NON-mixed case file system easily.  Linux has this feature, but setting up the file system is not easy.  You would need to split storage for linux containers from Windows containers, onto two separate partitions, and no one has worked on it.  Then getting an OCI Runtime with crun wrapping WINE.  Bottom line is the core maintainers have not worked on this, mainly because of time. If outside contributors worked on it, it would move ahead faster.

**Comment by Hi-Angel** [comment]
> this issue had no activity for 30 days

Here's a ⁺¹ comment to create activity. That could be another big advantage of Podman over Docker.

**Comment by a-clean-kitchen** [comment]
Commenting to show interest 👍 and keep this issue alive!

**Comment by rhatdan** [comment]
There is discussion on this happening in crun.
https://github.com/containers/crun/issues/1178

**Comment by wbehrens-on-gh** [comment]
Commenting to show interest

**Comment by fb314159265** [comment]
Commenting to show interest

**Comment by bhack** [comment]
> I honestly do not see how we could support Windows Containers on Linux.  We do not have Windows expertise, and once merged we have to maintain it and give upstream support.
> 
> I really like the idea though and find it cool but I think that shipping such a feature is just not sustainable for upstream.

True, but have you explored a maintaiener partnership with the Wine ecosystem?

**Comment by rhatdan** [comment]
I agree we don't have people to support this in the RHEL sense, but if we could get Wine experts to contribute and work with this, I think it would be a nice Upstream feature for Podman.

**Comment by vrothberg** [comment]
My comment above focused on upstream only.  I don't see how we can even do community support without any expertise.  I personally prefer to not ship a feature upstream rather than give the impression that incoming issues can be triaged or get any priority.

**Comment by seanyoung** [comment]
I've spent a little time looking at how windows containers work. There are two modes Windows containers can be run in, using the windows docker command line tools it looks like this:

```
docker container run --rm -it --isolation=hyperv mcr.microsoft.com/windows/server:ltsc2022
```
or 
```
docker container run --rm -it --isolation=process mcr.microsoft.com/windows/server:ltsc2022
```
In process mode, the container runs like you would expect: the windows kernel services system calls from the container directly. Emulating this mode requires implementing Windows system calls or using wine. 

The Hyper-V mode is interesting. There a small "utility VM" which runs in Hyper-V (a hypervisor). In the docker container there is a `UtilityVM` directory that contains a bunch of UEFI files that make up an OS. If I am reading the docker source code directly, both the UtilityVM and the docker container files are presented to the VM as a vhdx file. I guess we would need to figure out how to boot this VM and provide it with the disks it needs (I doubt that uses virtio based disks!).


https://forums.docker.com/t/use-of-container-utility-vm-folder-docker/118403/6
https://learn.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/hyperv-container
https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-11

**Comment by bhack** [comment]
So does it means that it could theoretically boot with qemu hyperv emulation?
https://www.qemu.org/docs/master/system/i386/hyperv.html

**Comment by seanyoung** [comment]
> So does it means that it could theoretically boot with qemu hyperv emulation? https://www.qemu.org/docs/master/system/i386/hyperv.html

I suspect so. Probably require some debugging to figure out exactly how to set things up.

**Comment by JohnnyElvis** [comment]
Are there any topics where a person without the ability to read code could help?

**Comment by Hi-Angel** [comment]
It seems, implementing Windows containers could fill a market with no competitors. Quoting a bit [of discussion from kde-devel](https://mail.kde.org/pipermail/kde-devel/2024-February/002461.html)

> For Windows however, Microsoft has limited the container stack to not allow
> anything GUI related to work. The underlying libraries may be there, but
> the equivalent display server components are not operational.
>
> To complicate things further, on Windows certain permissions are restricted
> to the interactive console and are not possible to do as either a scheduled
> task or as a system service.
> Usage of existing Windows automation frameworks such as Powershell Remoting
> or SSH will therefore not work if we want things to perfectly replicate a
> end user environment - because those will run the command(s) as part of a
> non-interactive session (even if the user we connect as is the same one
> logged in on the desktop console).

If Podman could run Windows containers, there would be no reason to adhere to such arbitrary limitations. Which would make it the preferred *(the only one really)* choice for a number of tasks.

**Comment by vrapolinario** [comment]
Hello folks, MSFT person here - first, wanted to make myself available to answer questions on Windows containers. Keep in mind I have close to zero knowledge on Linux. I focus on Windows containers, Kubernetes, Docker, etc. but everything running on the Windows OS (aside from the needed K8s control plane stuff).

I understand the desire to run Windows containers on Linux, so let me just put something out there: While that might be possible with Wine (or whatever method) that would not be supported by Microsoft. Of course, something working and being supported are different things and if this works, I'm sure some users will end up using it - I'm more just making this clear than anything else.

The other thing I wanted to understand is: Wouldn't be easier to actually have Podman support Windows containers on Windows? Today, docker, moby, and containerd interact with Windows (for Windows containers) via HCS. https://learn.microsoft.com/en-us/virtualization/api/hcs/overview 
I might be wrong, but I assume there could be some translation of container operation for Linux to HCS. I'm not saying it would be easy, but most likely easier than the running Windows containers on Linux? Maybe even take this discussion in parallel?

**Comment by Hi-Angel** [comment]
> The other thing I wanted to understand is: Wouldn't be easier to actually have Podman support Windows containers on Windows?

Isn't it a supported case?

That said, I'm not sure I get your point: do you want a Linux running a Windows VM running a Windows container or how do you want that to work on Linux? And even if so, that still requires buying a license, does it not…? I don't see how that would solve the case.

**Comment by 1480c1** [comment]
Seems the idea of windows containers on windows was mentioned before https://github.com/containers/podman/issues/11809

**Comment by vrapolinario** [comment]
@1480c1 Thanks for noting that. That's unfortunate. I could see huge advantages to having Podman support for Windows containers.

**Comment by Hi-Angel** [comment]
> > do you want a Linux running a Windows VM running a Windows container or how do you want that to work on Linux?
> 
> Well, the idea would be for Podman to run on Windows - if that's on a Windows host directly or Linux running a Windows VM, it wouldn't make difference.

I think there's a confusion. The issue is about Linux host to run Windows container. E.g. for CI purposes or for testing purposes, or just to run some service… So I am trying to figure out how your suggestion fits into a solution, and running a whole VM so far the only one I see. If that's indeed your suggestion, then needless to say that sounds very heavyweight.
 
> > And even if so, that still requires buying a license, does it not…? I don't see how that would solve the case.
> 
> Well, Windows containers need to be license anyway. If you manage to run a Windows container on Linux machine, you are using a Windows OS, so that would have to be licensed anyway.

No, people have been running Windows apps on Mac OS and Linux via WINE with no license required for ages. Admittedly, Idk what a Windows container contains, but I've never heard of anyone requiring to pay to launch a container.

**Comment by vrapolinario** [comment]
> I think there's a confusion. The issue is about Linux host to run Windows container. 

No, I get that. I was proposing a different approach, which based on the comment above has been rejected. I was proposing that Podman could run Windows containers on Windows - they can still be in a CI/CD pipeline, just a different OS. I get this is not the original proposal, but based on the comments I felt this was becoming a harder task than anticipated, hence my suggestion.

> No, people have been running Windows apps on Mac OS and Linux via WINE with no license required for ages. Admittedly, Idk what a Windows container contains, but I've never heard of anyone requiring to pay to launch a container.

That's the thing. If you get an MSI/EXE package that was compiled for Windows and run on Linux, (or MacOS) that's totally fine. You are not using the Windows binaries, but simply translating the APIs. Now, if you run a Windows container, then you have Windows binaries, which need to be licensed. Don't shoot the messenger 😊.

**Comment by Frankkkkk** [comment]
> For more information on licensing for Windows containers: https://learn.microsoft.com/en-us/virtualization/windowscontainers/images-eula

That's for the windows base image. Not all windows containers use these base images, isn't that so ?

**Comment by vrapolinario** [comment]
Yes, they do. All Windows images use the Windows containers base images from MCR. You might have to track the original base image if you are using .Net for example, but they all do use the base images. If it is a Windows containers, there's no escape from that.

**Comment by Hi-Angel** [comment]
> I was proposing that Podman could run Windows containers on Windows - they can still be in a CI/CD pipeline, just a different OS.

This "different OS" isn't hackable *(due lack of source code)*, which isn't something developers are typically interested in. So I don't see your suggestion as a viable solution. Even if what you say about the license is true, well… at least the amount of "non-debuggable opaque code" would be reduced as much as possible when there's no actual Windows host around the container.

**Comment by Jackenmen** [comment]
@vrapolinario If there's interest from Microsoft to champion the effort of Podman support for Windows containers on Windows then the relevant issue is #11809 and since it's locked, a new one would have to be created. The conclusion on that issue was that the Podman community needs Windows engineers to allow Podman to support this. Making yourself available to help with this is a good first step but I imagine it may not be enough unless some volunteer jumps in so MSFT may just need to help implement this themselves. Either way, a new issue mentioning #11809 would be a good way to move this forward since this issue is about a different (likely harder to implement and unlikely to be of Microsoft's interest) feature of supporting Windows containers on Linux (where IMO the main benefit is that it requires A LOT less resources than running a full Windows VM with a Docker/Podman daemon that you can connect to from the Linux host it's running on).

**Comment by vrapolinario** [comment]
I can't (and won't) debate that. I didn't mean to stear off the original conversation. Was just proposing an alternative which would help Windows apps be used with Podman. Honestly, I'd love to be able to tell Microsoft and Windows customers that they have Podman as an option to run Windows containers.

In any case, I'm here if you have question on Windows containers and I can help. 

And please, keep in mind the licensing. It's not if it's true. If it is a Windows container it requires a Windows license.

**Comment by vrapolinario** [comment]
> @vrapolinario If there's interest from Microsoft to champion the effort of Podman support for Windows containers on Windows then the relevant issue is #11809 and since it's locked, a new one would have to be created. The conclusion on that issue was that the Podman community needs Windows engineers to allow Podman to support this. Making yourself available to help with this is a good first step but I imagine it may not be enough unless some volunteer jumps in so MSFT may just need to help implement this themselves. Either way, a new issue mentioning #11809 would be a good way to move this forward since this issue is about a different (likely harder to implement and unlikely to be of Microsoft's interest) feature of supporting Windows containers on Linux (where IMO the main benefit is that it requires A LOT less resources than running a full Windows VM with a Docker/Podman daemon that you can connect to from the Linux host it's running on).

Sounds good. I'll take this internally, but even if there is interest, it won't happen so soon.

**Comment by geoffreysmith** [comment]
I don’t remember my original intent in starting the question but our issue has actually become worse and/or changed but we run “atomic desktops” on development machines for a number of reasons which usually use has a development workflow of toolbox/distrobox for sandboxed dev then eventually on commit whether or not the build is ready to product gets sent to a build matrix where more or less it gets turned turned into a lightweight K8S container that way we’re building to prod always and release on git tags. We do bypass the kernel but not often usually for performance reasons. Docker has come a long way in running rootless and/or reasons we began using CRIO.

So we use our host for OSX/Windows and ssh into a big SSH dev instance where we spend most our day using VSCode or VIM over SSH to big Debian boxes. WSL won’t work as we have auditing software (ZScalar) that prevents loopback and we do sometimes need to modify the Linux kernel for performance reasons very specific to us… but in any case 99% of our day is on a giant server and it’s much easier to have our host machine with Teams/etc and get good driver support for things like power management that only a few people will do the Debian laptop and VM into Windows thing for work communication.

But the problem we have now might be different. While our use of multi cluster setups is to support internal business apps and deal with large data intake, workflow and light on-prem ML. The rest of the company isn’t probably isn’t even aware and just knows their data is in OneDrive and that Office 365 just works. We will plug into the graph API but that’s simply to update reports, etc. which can be done entirely in Linux.

But that’s to just give you an overview. Our problem now is that it is trivial to output the results of 15tb of data process from multiple sources and drop it into essentially a certain teams project folder or watch when files change, etc. PowerPoint is presenting different challenges. This is definitely an edge use case but we can take the OpenXML spec , extract it have a generator that creates go code then “correctly” create guidelines, placeholders, etc from a designer who might’ve taken shortcuts. Then we have a content db where we store assets like logos and variables when we “know” certain things about the project. We found creating things from master templates/layouts/placeholders and not avoiding them even if we programmatically enter data avoids a lot of issues. We give then have a word document (user preference) where they enter data where we can’t fill out and essentially it creates a json schema we can validate and enter into text boxes etc. The point being is that PowerPoint does things like vertical-spacing poorly compared when you have text titles that might overflow to three lines to stay within a column layout and the description box will not adjust to 0.5” from the bottom. Same with grid lines where we can snap to but the user won’t or forget that 5 minutes before a meeting. We are trying to treat this essentially like a website with components that are aware of other components and adjust accordingly. Even worse is grouping or like adding a slide with project team members. Adding one at the last minute that can’t be validated by the server means we need a few simple plugins. Plus we also have the ability to have them mark items as basically “I want to not to stick to the design guide.”

We’d really like the ability to test this like we do Linux microservices. Create reasonable build matrix of Windows versions and PowerPoint versions and create test slides. Our syntax is usually correct but typography on Windows and other display issues means we need real copies, create a few test slides and make sure PowerPoint or whatever is looking right. Then run through them and basically create a selenium/web screenshot type setup so we can make sure we aren’t doing any regression or if a designer gives us a new slide what worked on their machine also works on lower DPI screens, different versions etc. what we are doing is simple and our validating the schema is correct we’d like to boot up a lightweight container, open a test file go through the slides take screenshots and make sure we past tests. Accidentally moving a dynamically placed text box off grid isn’t the end of the world and we aren’t looking for pixel  perfect. But right now we’d have to build ansible VMs of giant Windows images in HDR, several different resolutions last x versions and then multiply that by the last couple versions of PowerPoints. 

Containers let us build caches and we can often debloat them. We are a large >50k org and have whatever high level of partnership so licensing or whatever isn’t an issue I’m sure. It’s a pretty simple ask in the end “do our plugins or server side processing doing the same thing work” and while we do this with Linux microservices on every build there’s a huge difference between a 300 mb image with build matrix and cache is huge. Not to mention we use GitHub on-prem to push images up with the runner and some minor custom tooling and it “just works” with GitOps. We don’t need this to create an image though catching an error and rolling back the services and plugin package that power this would be nice.

So oddly while we do a lot of complex things we can call the Office365 API to make sure more or less the API didn’t change (some traditional sysadmin or secops person didn’t change permissions randomly), there’s no visual or Windows element to it so far as needing to not boot Windows.

What I’d rather not do is create some strange VM workflow maybe with snapshots to speed it up and create an entirely new runner, etc for one sort of minor thing.

So really any container that could slim Windows down and/or containerize it would be helpful. Not to kiss ass but the open nature of Azure/Office 365 (even though we do our K8S on-prem means we build out our apps and call out to a simple API means we actually kept to Microsoft).

Surely there’s a way internally manual apps are tested for graphical regression. I get that might not be available to us but the UI rendering in Windows differs so much we can’t reliably test it and I’m sure it’s due to years of bowl level crust built up over the years. Even if we don’t have a traditional UI interface and could only take screen shots (using ansible or some scripting to open a file, create a screenshot, move to next template, etc) we can create a Windows host cluster and deploy containers on them. Maybe calling DirectX (or whatever) but I’m not familiar enough with creating graphical apps to know how the rendering engine works. If we generate a UI it’ll be a website but we are mostly Rust/Golang/C++ programmers, so as long as we can do something to record output from the CLI that’s what we are looking for.

Sorry for the wall of text and maybe getting off topic but essentially we’re looking to generate screenshots from a container (docker or otherwise) even if it means layering on packages not included in the Windows container, Podman or Docker.

I’m sure internally you might literally have bare metal racks that not only have full Windows installs but also different physical graphics cards. But also your day to day devs probably have a slimmed down solution that works 90% of the time. 

Sorry for the wall of text but thinking this through a bit but I understand not wanting an interactive GUI for licensing issues on containers, but even if we have to call low level com interops to open up an app, open a file, move through slides, take screenshots that might actually be more automated and a better solution.

Let me know if you have any ideas. Thanks!

**Comment by branpurn** [comment]
> Sorry for the wall of text but thinking this through a bit but I understand not wanting an interactive GUI for licensing issues on containers, but even if we have to call low level com interops to open up an app, open a file, move through slides, take screenshots that might actually be more automated and a better solution.
> 
> Let me know if you have any ideas. Thanks!

Had Windows Server Nano not become vaporware, it was probably intended to evolve for use cases such as these

**Comment by tppalani** [comment]
Hi all

this issue opened very long back just want to know how to build windows based docker images using podman build command, is that podman will support?

**Comment by gmankab** [comment]
it would be so cool to see this feature in podman

**Comment by spamkutu** [comment]
Well, when are we making progress?

**Comment by ohault** [comment]
I can see two paths towards running Windows containers on Linux:
A) on Linux, running a hosting Windows VM in a similar way a Linux Moby VM is running on Windows to host and run Linux containers to provide a kind of Hyper-V isolation. 
B) using WINE to run Windows Containers.

**Comment by ntninja** [comment]
I downloaded one of the Windows containers using `skopeo` and aside from the weird hard-linking issues (which are probably not-too-hard to fix) and using custom PAX headers to represent some Windows permission information (which I think WINE supports mapping to user XAttrs) the container images are indeed complete Windows file systems including the kernel and bootloader (which we have learned are only for Hyper-V mode).

Of course, WINE does not actually support launching applications from an unmodified Windows installation: WINE cannot read Windows registry hives, WINE only implements a subset of Windows system calls that badly-behaving Windows applications directly invoke (mostly as part of copy-protection schemes, MSFT explicitly forbids doing this but they of course burdened themselves with infinite-compatibility so they still have to support it…). WINE using an unmodified Windows installation never has and never will be supported by WINE, also because of legal reasons.[^1]

This means that the only way to could be implemented both from a legal *and* a technical standpoint is to entirely forego downloading *any* base images from `mcr.microsoft.com` for the *process* isolation mode and instead patch all Windows Container images to instead use a pre-made WINE image containing the files as provided inside a `$WINEPREFIX` (+PowerShell which the Windows Container images also include, but it’s OpenSource) as a base. The level of “Windows-compatibility” of such an installation would of course correspond to that of the given version of WINE.

In order to avoid the whole static-linking problem of WINE, a simpler solution is to just include the WINE launcher binary and `wineserver` (the userland NT kernel emulation) as, say, `/WINE/*` and then launch every Docker command using `env WINEPREFIX=/ /WINE/bin/wine "${ORIG_CMD[@]}"` with the `/` directory containing what is usually `/drive_c` in a `$WINEPREFIX` as `/Files` (for Windows Image compatibilty) and the default registry hives files and the `/dosdevices` directory with `/dosdevices/c:` as a symlink to `../Files`. That way the WINE command can then launch `wineserver`, `services.exe`, `winedevice.exe`, `svchost.exe`, … as normal.

Note that WINE *does* supporting running headlessly (=without GUI support), but will fail window context-creation if no display server is available. I was not able to quickly find out if this matches the Windows Containers behaviour or if GUI windows may be created there but just cannot be accessed externally. To emulate that behaviour `Xvfb` would have to be included in the container images and launched as `/WINE/bin/Xvfb :0` before running the main command with `DISPLAY=:0` set.

Finally, emulating a case-insensitive file system at runtime is *not* required. WINELib and the `wineserver` can handle that just fine like they do on other Linux systems and will in fact to *better* than the Linux kernel since they use the ancient Unicode 1.1 NT filesystem path translation tables rather than modern versions of Unicode that (by necessity) include extended case-folding rules that won’t match those of Windows NT 3.5… What is needed though is implementing the WINE/Windows case-folding algorithm in when *extracting* images only.

 [^1]: Of course, *technically* if you replace the `user32.dll`, `kernel32.dll` and `ntdll.dll` with their WINE counterparts things still *might* work to an astonishing degree… :wink:

**Comment by slonopotamus** [comment]
> the container images are indeed complete Windows file systems **including the kernel and bootloader** (which we have learned are only for Hyper-V mode).

Is it enough to run a VM? Maybe Linux should just use the same route as Hyper-V isolation mode.

**Comment by seanyoung** [comment]
> > the container images are indeed complete Windows file systems **including the kernel and bootloader** (which we have learned are only for Hyper-V mode).
> 
> Is it enough to run a VM? Maybe Linux should just use the same route as Hyper-V isolation mode.

Exactly. The root to go is the boot the included kernel in a VM which supports Hyper-V and all the extensions. 

docker for windows supports windows containers on Windows. The first step might be to get podman to the same level of support on Windows first. Then we understand more about which windows API calls are being used, we can then try to make it work on Linux.

**Comment by TheMDev** [comment]
Microsoft was working on [Linux Hyper-V DOM0](https://lore.kernel.org/lkml/1727985064-18362-1-git-send-email-nunodasneves@linux.microsoft.com/) support at some point. That was the route I was hoping to explore if it ever happens.

**Comment by ohault** [comment]
> Microsoft was working on [Linux Hyper-V DOM0](https://lore.kernel.org/lkml/1727985064-18362-1-git-send-email-nunodasneves@linux.microsoft.com/) support at some point. That was the route I was hoping to explore if it ever happens.

I would be please to read a blog article about this with a comprehensive step by step guide about how to test these new Linux kernel features.

**Comment by TheMDev** [comment]
@ohault as far as I am aware none of the stuff they have implemented thus far is functional and I have not seen any updates on the topic in a couple years.

**Comment by ohault** [comment]
> @ohault as far as I am aware none of the stuff they have implemented thus far is functional and I have not seen any updates on the topic in a couple years.

The best I know so far is this project https://github.com/olljanat-ai/hyperv-for-linux

**Comment by ntninja** [comment]
> Is it enough to run a VM? Maybe Linux should just use the same route as Hyper-V isolation mode.

It *should* be enough, but we’d need to figure out how to exactly shape the OCI “Image” which is just a TAR file into whatever shape QEMU expects, it should then *probably* be possible to use QEMU’s Hyper-V support:

https://www.qemu.org/docs/master/system/i386/hyperv.html (it is actually for x86-64, contrary to the URL)

It might also require some undocumented Hyper-V features, then that would have to be reverse-engineered.

Reminder though, that is is definitely a violation of
https://learn.microsoft.com/en-us/virtualization/windowscontainers/images-eula
which requires the Container image to execute on a corresponding “Host Software”:

> You may use the Container Image **to create, build, and run** your Consolidated Image **on Host Software and** distribute the Container Image only as part of your Consolidated Image.

With the “Host Software” being defined as:

> Windows Server Standard or Windows Server Datacenter software (collectively “Server Host Software”), or Microsoft Windows Operating System (version 10) software (“Client Host Software”), or Windows 10 IoT Enterprise and Windows 10 IoT Core (collectively “IoT Host Software”). The Server Host Software, Client Host Software, and IoT Host Software are collectively referred to as the “Host Software” and a license for Host Software is a "Host License".

Since Linux is not in the above list, any commercial of such feature would definitely be illegal. Private use would probably be a “Fair Use”-ish thing if you have a “Host License” (which then can be whatever your last Windows-infested computer came with) though, so should be OK.

I just described the variant that would not require any Microsoft license whatsoever.

**Comment by ntninja** [comment]
> docker for windows supports windows containers on Windows. The first step might be to get podman to the same level of support on Windows first. Then we understand more about which windows API calls are being used, we can then try to make it work on Linux.

That would help with figuring out how to generate the image format, but not how to actually boot the result VM since the Windows APIs handle that internally:
https://github.com/podman-container-tools/podman/issues/8136#issuecomment-1733555366

The HCS interface looks a lot like libvirt to me, honestly:
https://learn.microsoft.com/en-us/virtualization/api/hcs/overview

The way to go here, is really to read as much information from the Docker on Windows source code and:

 1. Document how it downloads and unpacks the Windows OCI images
 2. Builds the disk image
 3. Invokes HCS to launch the VM

Step 1 is also relevant for swapping in WINE with process isolation (https://github.com/podman-container-tools/podman/issues/8136#issuecomment-5273404354), step 2 should result in actionable code ideally, step 3 is hopefully enough for fiddling with QEMU to get a compatible environment based on whatever Docker for Windows does.

On Windows (when considering Podman on Windows), steps 1 and 2 should be basically the same other than storing metadata the Windows way during unpack. Step 3 would obviously map directly to the native HCS calls.

**Comment by TheMDev** [comment]
containerd and buildkit also have a good amount of information on wcow support as well
